### ISSUE TITLE
Upgrade Rust Crytpo asymmetric dependencies

### ISSUE THREAD

**Issue by agostbiro** [summa-tx/bitcoins-rs#96]
I'm trying to upgrade dependencies in [ethers-rs](https://github.com/gakonst/ethers-rs) and came across a problem where dependency conflicts with `coins-bip32` are blocking the upgrade. I've tried to upgrade relevant packages in `coins-bip32`, but can't seem to get one of the macros right. I'm adding more detail in the attached PR.

Diff in ethers-rs:

```diff
diff --git a/ethers-core/Cargo.toml b/ethers-core/Cargo.toml
index 00fa752..ae96486 100644
--- a/ethers-core/Cargo.toml
+++ b/ethers-core/Cargo.toml
@@ -16,9 +16,9 @@ arrayvec = { version = "0.7.2", default-features = false }
 rlp-derive = { version = "0.1.0", default-features = false }
 -elliptic-curve = { version = "0.11.12", default-features = false }
+elliptic-curve = { version = "0.12.1", default-features = false }
-k256 = { version = "0.10.4", default-features = false, features = ["keccak256", "ecdsa", "std"] }
+k256 = { version = "0.11.2", default-features = false, features = ["keccak256", "ecdsa", "std"] }
```

Conflict in ethers-rs:

```
error: failed to select a version for `signature`.
    ... required by package `ecdsa v0.13.3`
    ... which satisfies dependency `ecdsa-core = "^0.13"` of package `k256 v0.10.0`
    ... which satisfies dependency `k256 = "^0.10"` of package `coins-bip32 v0.6.0`
    ... which satisfies dependency `coins-bip32 = "^0.6.0"` of package `coins-bip39 v0.6.0`
    ... which satisfies dependency `coins-bip39 = "^0.6.0"` of package `ethers-signers v0.13.0 (/Users/abiro/repos/ethers-rs/ethers-signers)`
    ... which satisfies path dependency `ethers-signers` (locked to 0.13.0) of package `ethers v0.13.0 (/Users/abiro/repos/ethers-rs)`
    ... which satisfies path dependency `ethers` (locked to 0.13.0) of package `ethers-wasm v0.1.0 (/Users/abiro/repos/ethers-rs/examples/ethers-wasm)`
versions that meet the requirements `>=1.3.1, <1.5` are: 1.4.0, 1.3.2, 1.3.1

all possible versions conflict with previously selected packages.

  previously selected package `signature v1.5.0`
    ... which satisfies dependency `signature = "^1.5"` of package `ecdsa v0.14.2`
    ... which satisfies dependency `ecdsa-core = "^0.14"` of package `k256 v0.11.2`
    ... which satisfies dependency `k256 = "^0.11.2"` of package `ethers-core v0.13.0 (/Users/abiro/repos/ethers-rs/ethers-core)`
    ... which satisfies path dependency `ethers-core` (locked to 0.13.0) of package `ethers v0.13.0 (/Users/abiro/repos/ethers-rs)`
    ... which satisfies path dependency `ethers` (locked to 0.13.0) of package `ethers-wasm v0.1.0 (/Users/abiro/repos/ethers-rs/examples/ethers-wasm)`

failed to select a version for `signature` which could resolve this conflict
```

**Comment by agostbiro** [comment]
I've managed to upgrade the `coins-bip32` crate to `k256 = "0.11"` and `digest = "0.10.3"`, but I'm running into trouble upgrading the hashers in `coins-core` to satisfy the new trait bounds required by [`DigestSigner`](https://docs.rs/k256/latest/k256/ecdsa/struct.SigningKey.html#impl-DigestSigner%3CD%2C%20Signature%3CSecp256k1%3E%3E) in `k256 = "0.11"`. There are two blockers:

### 1. Potential panic

[`digest::VariableOutput::finalize_variable`](https://docs.rs/digest/latest/digest/trait.VariableOutput.html#tymethod.finalize_variable) now returns a `Result`, but [`digest::FixedOutput`](https://docs.rs/digest/latest/digest/trait.FixedOutput.html) methods are infallible, which means we have to unwrap the `Result` and potentially panic.

Original code:

```rust
impl digest::FixedOutput for Blake2b256 {
    fn finalize_into(self, out: &mut DigestOutput<Self>) {
        digest::VariableOutput::finalize_variable(self.0, |res| {
            AsMut::<[u8]>::as_mut(out).copy_from_slice(&res[..32])
        });
    }
}
```

Code with `digest = "0.10.3"`:

```rust
impl digest::FixedOutput for Blake2b256 {
    fn finalize_into(self, out: &mut DigestOutput<Self>) {
        digest::VariableOutput::finalize_variable(self.0, out).unwrap();
    }
}
```

I don't understand well the uses of `Blake2b256` in the project, but I'm wondering if it'd be possible to use a fixed output variant as the inner hasher which would avoid sidestepping this issue.

### 2. Unsatisfied trait bounds

[`DigestSigner`](https://docs.rs/k256/latest/k256/ecdsa/struct.SigningKey.html#impl-DigestSigner%3CD%2C%20Signature%3CSecp256k1%3E%3E) now requires the [`digest::core_api::CoreProxy`](https://docs.rs/digest/0.10.3/digest/core_api/trait.CoreProxy.html), which in turn requires the private `digest::core_api::wrapper::sealed::Sealed` trait. I've opened an [issue](https://github.com/RustCrypto/traits/issues/1028) in the `digest` repo to resolve this.

**Comment by agostbiro** [comment]
Thanks for #101 @mattsse ! I'm a little bit concerned about these lines:

https://github.com/summa-tx/bitcoins-rs/blob/a1739f867e3fd6bef9cb170d6ed1e041a719db7c/core/src/hashes/mod.rs#L218

https://github.com/summa-tx/bitcoins-rs/blob/a1739f867e3fd6bef9cb170d6ed1e041a719db7c/core/src/hashes/mod.rs#L228

If I'm not mistaken a potential error is being ignored there as `digest::VariableOutput::finalize_variable` now returns a `Result`.

--

Btw we've relaxed the `DigestSigner` bounds in the `digest` crate in-between so there is potential for simplification in the hash implementations in the future:

https://github.com/RustCrypto/elliptic-curves/pull/613/files

**Comment by prestwich** [comment]
According to the [docs for variable output](https://docs.rs/digest/latest/digest/trait.VariableOutput.html#tymethod.finalize_variable), that method errors if the buffer is of incorrect size, which we know not to be the case

**Comment by agostbiro** [comment]
Btw I've noticed that the `DigestSigner` update has been released in `k256 = 0.11.3` and the PR is actually using that version already, so I'm closing this issue.

**Comment by mattsse** [comment]
noticed I didn't delete the old impl docs...
used `expect` rather than silently dropping the result https://github.com/summa-tx/bitcoins-rs/pull/103
